mcs/class/System/System.Net.Sockets/Socket.cs

```diff
@@ -972,7 +973,32 @@ public IAsyncResult BeginConnect (string host, int port, AsyncCallback requestCa
 			if (is_listening)
 				throw new InvalidOperationException ();
 
-			return BeginConnect (Dns.GetHostAddresses (host), port, requestCallback, state);
+			var callback = new AsyncCallback ((result) => {
+				IPAddress[] hostAddresses = ((Task<IPAddress[]>)result).Result;
+				BeginConnect (hostAddresses, port, requestCallback, state);
```

Comment:
to avoid capture, maybe just use `((Task<IPAddress[]>)result).AsyncState`

Reply:
btw, this will still capture the context due to port and requestCallback parameters